File: src/main/java/me/creepinson/creepinoutils/util/GroupTextComponent.java
package me.creepinson.creepinoutils.util;

import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.StringTextComponent;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.util.text.TranslationTextComponent;

import javax.annotation.Nonnull;

/**
 * Originally created by Thiakil.
 * This is a utility class to make text components fluent.
 * @author Theo Paris
 */
public class GroupTextComponent extends StringTextComponent {

    public GroupTextComponent() {
        super("");
    }

    public GroupTextComponent(TextFormatting color) {
        super("");
        getStyle().setColor(color);
    }

    @Nonnull
    @Override
    public String getUnformattedComponentText() {
        return "";
    }

    @Nonnull
    @Override
    public GroupTextComponent shallowCopy() {
        GroupTextComponent StringTextComponent = new GroupTextComponent();
        StringTextComponent.setStyle(this.getStyle().createShallowCopy());
        for (ITextComponent itextcomponent : this.getSiblings()) {
            StringTextComponent.appendSibling(itextcomponent.shallowCopy());
        }
        return StringTextComponent;
    }

    public GroupTextComponent string(String s) {
        this.appendSibling(new StringTextComponent(s));
        return this;
    }

    public GroupTextComponent string(String s, TextFormatting color) {
        ITextComponent t = new StringTextComponent(s);
        t.getStyle().setColor(color);
        this.appendSibling(t);
        return this;
    }

    public GroupTextComponent translation(String key) {
        this.appendSibling(new TranslationTextComponent(key));
        return this;
    }

    public GroupTextComponent translation(String key, Object... args) {
        this.appendSibling(new TranslationTextComponent(key, args));
        return this;
    }

    public GroupTextComponent translation(String key, TextFormatting color) {
        ITextComponent t = new TranslationTextComponent(key);
        t.getStyle().setColor(color);
        this.appendSibling(t);
        return this;
    }

    public GroupTextComponent translation(String key, TextFormatting color, Object... args) {
        ITextComponent t = new TranslationTextComponent(key, args);
        t.getStyle().setColor(color);
        this.appendSibling(t);
        return this;
    }

    public GroupTextComponent component(ITextComponent component) {
        this.appendSibling(component);
        return this;
    }

    public GroupTextComponent component(ITextComponent component, TextFormatting color) {
        component.getStyle().setColor(color);
        this.appendSibling(component);
        return this;
    }
}

File: src/main/kotlin/com/theoparis/creepinoutils/util/text/GroupTextComponent.kt
/*
 * Copyright (c) Creepinson
 */
package com.theoparis.creepinoutils.util.text

import net.minecraft.util.text.*
import javax.annotation.Nonnull

/**
 * Originally created by Thiakil.
 * This is a utility class to make text components fluent.
 *
 * @author Theo Paris
 */
class GroupTextComponent : StringTextComponent {
    constructor() : super("")
    constructor(color: TextFormatting) : super("") {
        style.color = Color.fromTextFormatting(color)
    }

    @Nonnull
    override fun getUnformattedComponentText(): String {
        return ""
    }

    override fun copyRaw(): GroupTextComponent {
        val stc = GroupTextComponent()
        stc.style = style
        for (itextcomponent in getSiblings()) {
            stc.append(itextcomponent.deepCopy())
        }
        return stc
    }

    fun string(s: String): GroupTextComponent {
        this.append(StringTextComponent(s))
        return this
    }

    fun string(s: String, color: TextFormatting): GroupTextComponent {
        val t: ITextComponent = StringTextComponent(s)
        t.style.color = Color.fromTextFormatting(color)
        this.append(t)
        return this
    }

    fun translation(key: String): GroupTextComponent {
        this.append(TranslationTextComponent(key))
        return this
    }

    fun translation(key: String, vararg args: Any): GroupTextComponent {
        this.append(TranslationTextComponent(key, *args))
        return this
    }

    fun translation(key: String, color: TextFormatting): GroupTextComponent {
        val t: ITextComponent = TranslationTextComponent(key)
        t.style.color = Color.fromTextFormatting(color)
        this.append(t)
        return this
    }

    fun translation(key: String, color: TextFormatting, vararg args: Any): GroupTextComponent {
        val t: ITextComponent = TranslationTextComponent(key, *args)
        t.style.color = Color.fromTextFormatting(color)
        this.append(t)
        return this
    }

    fun component(component: ITextComponent): GroupTextComponent {
        this.append(component)
        return this
    }

    fun component(component: ITextComponent, color: TextFormatting): GroupTextComponent {
        component.style.color = Color.fromTextFormatting(color)
        this.append(component)
        return this
    }
}